### PR TITLE
[WIP] [jsk_robot_utils] Add robot-interface-utils.l to resolve name and file where  robot-interface is defined

### PR DESCRIPTION
**Pull request by garaemon**
```lisp
(load (robot-file (ros::get-param "/robot")))
(init-robot-from-name (ros::get-param "/robot"))
```

### CODE REVIEW

**Comment by garaemon** [comment]
* loading robot-interface file is heavy because it load robot model definition
  * :arrow_right: need to load one file rather than load all the files
* it is difficult to know where the robot-interface class is defined
  * :arrow_right: write down the relationship between name of robots and files
  * :arrow_right: brom a viewpoint of class hierarchy, writing `robot-file` function in `robot-interface.l` is not straight forward.

**Comment by garaemon** [comment]
Original definition is [here](https://github.com/jsk-ros-pkg/jsk_demos/blob/master/jsk_2015_06_hrp_drc/drc_task_common/euslisp/robot-util.l#L7)

**Comment by k-okada** [comment]
wehre do we define `init-robot-from-name`

is it something related to `make-robot-model-from-name` https://github.com/euslisp/jskeus/blob/master/irteus/irtutil.l#L378

do we still need xx-init function? https://github.com/start-jsk/rtmros_tutorials/blob/master/hrpsys_ros_bridge_tutorials/euslisp/hrp2jsk-interface.l#L44

**Comment by garaemon** [comment]
`init-robot-from-name` is defined in [robot-interface.l](https://github.com/jsk-ros-pkg/jsk_pr2eus/blob/master/pr2eus/robot-interface.l#L1077)

**Comment by garaemon** [comment]
`pr2-init` has small hook in [pr2-init](https://github.com/jsk-ros-pkg/jsk_pr2eus/blob/master/pr2eus/pr2-interface.l#L802) but I think we can remove the hooks and just use `(instance pr2-interface :init)`

**Comment by garaemon** [comment]
`make-robot-model-from-name` and `make-robot-interface-from-name` is a simple function to create
robot model and robot interface from string rather than class object.

User should load proper model files before using the functions.

**Comment by garaemon** [comment]
We have several alternatives to archive this goal:
1. write `robot-file` function in `robot-interface.l`
2. write `robot-file` function in somewhere else as this PR does
3. All the robot should have `${ROBOT_NAME}eus` package like pr2eus and peppereus and programs can load proper `interface.l` file.

**Comment by k-okada** [comment]
so how about moving `make-robot-model-from-name` and `make-robot-interface-from-name` to robot-interface-utils.l so we can understand how  to use them

Another question is : are you suggesting people to use
```
(load (robot-file (ros::get-param "/robot")))
(init-robot-from-name (ros::get-param "/robot"))
(make-robot-model-from-name)
(make-robot-interface-from-name)
```
instead of calling `(hrp2-init)` ?

File: jsk_robot_common/jsk_robot_utils/euslisp/robot-interface-utils.l [hunk 1]
```diff
@@ -0,0 +1,32 @@
+(defun robot-file (name)
+  "You can get robot-interface file according to `name' argument.
+You can create `*ri*' like 
+(progn (load (robot-file (ros::get-param \"/robot\")))
```
Comment (k-okada) [inline jsk_robot_common/jsk_robot_utils/euslisp/robot-interface-utils.l]: please follow https://github.com/ros-infrastructure/rep/pull/104/files and use standard param name for robot type, currently `robot/type` is suggested.

**Comment by garaemon** [comment]
> Another question is : are you suggesting people to use
> 
> (load (robot-file (ros::get-param "/robot")))
> (init-robot-from-name (ros::get-param "/robot"))
> (make-robot-model-from-name)
> (make-robot-interface-from-name)
> instead of calling (hrp2-init) ?

My recommendation is:
```lisp
(load (robot-file (ros::get-param "/robot")))
(init-robot-from-name (ros::get-param "/robot"))
```

rather than
```lisp
(load "package://hrpsys_ros_bridge_tutorials/euslisp/hrp2jsknt-interface.l")
(hrp2jsknt-init)
```

`make-robot-model-from-name` should be useful to the people who want to use multiple robot models in planning and so on.
I supposed `make-robot-interface-from-name` should be enough to create `*ri*` for each robot but
 it is a little bit convenient because it creates `*robot*P and `*ri*` at the same time.

**Comment by k-okada** [comment]
so can we remove
`(make-robot-model-from-name)` from https://github.com/euslisp/jskeus/blob/master/irteus/irtutil.l#L378
`(make-robot-interface-from-name) ` from 
https://github.com/jsk-ros-pkg/jsk_pr2eus/blob/master/pr2eus/robot-interface.l#L1070

and all `xx-init` function from https://github.com/start-jsk/rtmros_tutorials/blob/master/hrpsys_ros_bridge_tutorials/euslisp/hrp2jsk-interface.l#L44, https://github.com/start-jsk/rtmros_tutorials/blob/master/hrpsys_ros_bridge_tutorials/euslisp/hrp2jsknt-interface.l#L67, and https://github.com/start-jsk/rtmros_tutorials/blob/master/hrpsys_ros_bridge_tutorials/euslisp/jaxon-interface.l#L112 ... ?

**Comment by garaemon** [comment]
Yes,

**Comment by k-okada** [comment]
humm, that's big change and I do not think people will switch from 
```
hrp2-init
```
to
```
(load (robot-file (ros::get-param "/robot")))
(init-robot-from-name (ros::get-param "/robot"))
```

how about create 
```
robot-init
```
that do every thing for you.

but still I'm afraid people using their custom xx-init function like 
https://github.com/jsk-ros-pkg/jsk_pr2eus/commit/284901f4ee482eca69483452f0737f3d7c66e47b 
...



--
◉ Kei Okada

**Comment by garaemon** [comment]
> that's big change and I do not think people will switch from

The easiest solution is to remove `xx-init` functions.



> how about create
> ```
> robot-init
> ```
> that do every thing for you.

sounds good but robot-init conflicts with a lot of existing codes.

**Comment by garaemon** [comment]
> but still I'm afraid people using their custom xx-init function like
> https://github.com/jsk-ros-pkg/jsk_pr2eus/commit/284901f4ee482eca69483452f0737f3d7c66e47b

it was added 3 years ago and recently no robot has custom init function. 
However I know all the code requires special initialization for each task.

**Comment by k-okada** [comment]
So, question is most of people agree to use robot-init function instead of
using their xxx-init function. Anyone?

2015年8月3日月曜日、Ryohei Ueda<notifications@github.com>さんは書きました:

> but still I'm afraid people using their custom xx-init function like
> jsk-ros-pkg/jsk_pr2eus@284901f
> <https://github.com/jsk-ros-pkg/jsk_pr2eus/commit/284901f4ee482eca69483452f0737f3d7c66e47b>
>
> it was added 3 years ago and recently no robot has custom init function.
> However I know all the code requires special initialization for each task.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jsk-ros-pkg/jsk_robot/pull/401#issuecomment-127188070>
> .
>


-- 
--
◉ Kei Okada

**Comment by garaemon** [comment]
> So, question is most of people agree to use robot-init function instead of
> using their xxx-init function. Anyone?

any opinions?

**Comment by garaemon** [comment]
特に意見が無いようですが，`${robot}-init`を廃止して，全ロボット対応版の`robot-init`を導入して良いですか?

**Comment by snozawa** [comment]
（話についてくのが遅くなりすいません。。。）

`robot-init`をよんだときに、各ロボットを指定する方法はどうするかんじになりますでしょうか？
get-param？
また、robots.lのようなものをつくることになるでしょうか？

**Comment by garaemon** [comment]
```lisp
(defun robot-init (&optional (robot-name (ros::get-param "/robot/type" (unix::getenv "ROBOT")))
  ...)
```
みたいな感じじゃないでしょうか。
デフォルトは`/robot/type`で、なかったら$ROBOTになって、optionalで指定もできる。

robotの名前とファイルの対応付けはどこかに書かなくてはいけないですね。

**Comment by snozawa** [comment]
環境変数つかいますか？

**Comment by garaemon** [comment]
> 環境変数つかいますか？

個人的にはどちらでも良い気とおもっています。暗黙的な何かがはいってしまうからあえて環境変数から初期値はなくしたほうが良いかもしれません

**Comment by snozawa** [comment]
どうしましょうか。
使い方をいくつもサポートするのも大変なので、
　rosparamにセットしておく
　明示的に`robot-init "hrp2jsk"`などをよぶ
に絞っていくので良いきがします。

**Comment by snozawa** [comment]
robot-initで全部できるようにする
（loadもhogehoge-initもする）
に賛成で、ロボットを指定する方法は
- 環境変数はなし
- 引数であたえる（デフォルト引数に`ros::get-param`）

が良いのではないでしょうか。

**Comment by garaemon** [comment]
作業し始めてやはり`xx-init`を消して`robot-init`に強制的に移行させるというのはかなりラディカルな思想だなという気がしてきました。
warnを出すくらいの段階を考えるべきですね.

```lisp
(defun pr2-init ()
  (load ...)
  (warn "YOU NEED TO USE ROBOT-INIT FUNCTION~%")
  (warn "PR2-INIT IS DEPRECATED AND IT WILL BE REMOVED IN NEAR FEATURE~%")
  (robot-init "pr2"))
```

**Comment by snozawa** [comment]
> 作業し始めてやはりxx-initを消してrobot-initに強制的に移行させるというのはかなりラディカルな思想だなという気がしてきました。
> warnを出すくらいの段階を考えるべきですね.

あ、それで消してたんですね。
僕は
> warnを出すくらいの段階を考えるべきですね.

で移行期を設けるものかと思ってました。
移行期はあったほうがよいと思います。
できればwarning-messageでカラーになってるとオシャレ

**Comment by garaemon** [comment]
sedスクリプトくらい用意しないと大変そうです。
一回プル陸を消したりして変更を整理します。

**Comment by garaemon** [comment]
作業して気づきましたが、レポジトリをやはり変えるべきな気がしてきました。
robot-interface.lと同じレイヤで定義すべき問題な気がする。
ちょっと頭を冷やして落ち着いて整理したほうが良さそう。

**Comment by snozawa** [comment]
> robot-interface.lと同じレイヤで定義すべき問題な気がする。
> ちょっと頭を冷やして落ち着いて整理したほうが良さそう。

リンクをたどるだけだったので今まで気づきませんでしたが、ここjsk_robotだったんですね。。。
robot-interface.lのファイルも読み込むので、robot-interface.lにあるのが良い気がします。

**Comment by garaemon** [comment]
一旦closeして仕切り直します

**Comment by snozawa** [comment]
こちらの議論ですが，ロボット名をenvでなくて引数or rosparamでということだったと思いますが，
- 名前は`/robot/type`でOKでしょうか．個人的にはよさそうに思います．genericすぎという声もあるかもしれません．
- rosparam setするファイルはどのlaunchにしたら良いでしょうか．今JSKのロボット全員がincludeしてるlaunchファイルってあるんでしたっけ？

**Comment by k-okada** [comment]
> 名前は/robot/typeでOKでしょうか．個人的にはよさそうに思います．genericすぎという声もあるかもしれません

This has been proposed and discussed at here ->
https://github.com/ros-infrastructure/rep/pull/104

--
◉ Kei Okada

2015/12/25 18:35、Shunichi Nozawa <notifications@github.com> のメッセージ:
．

**Comment by snozawa** [comment]
I see.

**Comment by YoheiKakiuchi** [comment]
すごい蒸し返したはなしになるけど、環境変数はなんで無しなんだっけ？
rosparam

**Comment by snozawa** [comment]
おおもとの議論は
https://github.com/start-jsk/rtmros_hrp2/issues/237
https://github.com/start-jsk/rtmros_hrp2/issues/268
にありますが、環境変数でなくrosparamからとってくるメリットは
https://github.com/start-jsk/rtmros_hrp2/issues/268#issuecomment-126547508
の、例えばROSのマスタが一環して情報を配る、などがあります。

環境変数が必要な部分もありそうで、hrp2のc、trans_systemのビルド系があります。

実際どっちにしてくかはケースバイケースですが、最近目にするロボット名解決のケースでは以下がありそうです。
- launchにかいてあるものは、envでなくてargでよさそう。一番トップレベルでは`jaxon_hogehoge.launch`のように呼ぶことがおおいので、このトップレベルからargで渡すことができそう。https://github.com/jsk-ros-pkg/trans_system/pull/575
- Euslispでかつ`(ros::roseus "hoge")`してるようなファイルは、ロボット名はrosparamの`/robot/name`などをみるのでよさそう。ros::initしてるので。
- EuslispでROS関係ないところは、比較的抽象化されていて、概ね知ってる限りそもそもロボット環境変数も`/robot/name`もなくかけるところがほとんどのようにみえる。

**Comment by snozawa** [comment]
https://github.com/ros-infrastructure/rep/pull/104
で特に記載がなさそうでしたので質問なのですが、
- robot/typeのロボットtypeは、hrp2jsknt, jaxon_red, ...etcのように、ROSの小文字、ハイフンでなくアンダーバー、などの命名速に則っていたらOKでしょうか。
- robot/nameは何でしょうか。さっき@furuchichovからオフラインできいて、pr2はホスト名ではないとのことでしたが、例えばhrp2の17の場合、hrp2017、hrp2017c、hrp2017vのいずれが良いでしょうか。

**Comment by furushchev** [comment]
PR2では
robot/typeがpr2で、robot/nameがpr1012,pr1040などのロボットを一意に特定できるものになっています。
pr1012はpr1012,pr1012s,pr1012basestationの3つのサーバからなるので、必ずしもhostnameと一致していないです。

**Comment by snozawa** [comment]
>  pr1012はpr1012,pr1012s,pr1012basestationの3つのサーバからなるので、必ずしもhostnameと一致していないです。

一つはhostnameと一致してるやつがあるというかんじかな？

**Comment by YoheiKakiuchi** [comment]
> pr1012はpr1012,pr1012s,pr1012basestationの3つのサーバからなるので、必ずしもhostnameと一致してい> ないです。
> 一つはhostnameと一致してるやつがあるというかんじかな？

これは、どっちかというとhostnameの名付け規則の問題じゃないかな。
robot/typeはクラス名、robot/nameはインスタンス名で、それをどう使うかは、また別の問題。

**Comment by snozawa** [comment]
rtmros_commonなロボットで自動でセットされるように、
hrpsys_ros_bridge.launchで
robot/type
robot/name
をデフォルトで設定されるようにしようと思ったのですが、
前者はロボット名でだいたい決まるのでOKそうですが、
後者はやめたほうがいいでしょうか（やめたほうがいいきがしてきました）。

**Comment by snozawa** [comment]
rtmros_commonにPRしました
https://github.com/start-jsk/rtmros_common/pull/944

**Comment by k-okada** [comment]
通常はモデルファイルが違ったら，違うロボット（robot/type クラス名），とするのがやりやすいとは思います．
ただ，そうするとHRP2／Jaxonは全部違うロボットになっちゃうので，あまり意味がないのでクラス名＝{hrp2,
jaxon}みたいなのはしょうがない気はしますが．

--
◉ Kei Okada

**Comment by snozawa** [comment]
> 通常はモデルファイルが違ったら，違うロボット（robot/type クラス名），とするのがやりやすいとは思います．

そうですね。
https://github.com/start-jsk/rtmros_common/pull/944
でもモデルファイルがちがったら別なrobot/typeとしてみました。
HRP２の場合、/robot/type=hrp2jsk, hrp2jsknt, hrp2jsknts, hrp2wで
JAXONの場合、/robot/type=jaxon, jaxon_redになります。
これで、VRMLおよびhrpsys_ros_bridge_tutorials/euslisp/xx-interface.lのファイルの分け方と同じになります。

**Pull request closed without merging**